list_stored_sd_metadata API.

pradeepfn opened the pull request:
Summary:
1\ Certain checkpoint load use cases are not aware of the properties of the data/tensors they want to load.
2\ These usecases include data loader checkpoints, reading data for post processing (when the original model definition is not available).
3\ There, we have to use saved checkpoint  (metadata) as our source of truth.
4\ This RFC proposal exposes the checkpoint metadata using a public API.

In this proposal we expose the stored state-dict metadata  (minus associated storage/chunk metadata). 

Chunk/storage details should not be exposed to the users and is a impl detail of the storage writer/reader.

Test Plan:
UT.

Rollback Plan:

Differential Revision: D80231457




cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci

pradeepfn commented:
I will fix the lint issues before landing. Please review the code minus lint signals.

pradeepfn commented:
1\ Moved the code out of state_dict_loader.
2\ Moved to a new file. Could not re-use existing utils file due to circular deps issues.

saumishr commented on torch/distributed/checkpoint/list_stored_sd.py:
```diff
@@ -0,0 +1,32 @@
+
+# pyre-strict
+
+import os
```
Looks good. File name is too specific to a single API. Lets have a rather generic name. Something like metadata_utils.py?

pradeepfn replied:
Done!. PTAL

saumishr reviewed (approved):
Looks good! Please fix the lint errors before landing.

seemethere commented on test/distributed/checkpoint/test_list_stored_state_dict.py:
```diff
@@ -0,0 +1,63 @@
+            "bytes_data": io.BytesIO(b"TrainingEpoch:4"),
+        }
+
+        dist_cp.save(
+            state_dict=state_dict_to_save,
+            storage_writer=dist_cp.FileSystemWriter(path=CHECKPOINT_DIR),
+            planner=dist_cp.DefaultSavePlanner(),
+        )
```
Your git porcelain is failing since your test run didn't properly clean up its files after things run. We do this to ensure developers local directories don't fill up with files that could potentially be staged for committing.

You need to ensure this file gets deleted, or ideally use a tempfile.

See https://docs.python.org/3/library/tempfile.html

pradeepfn committed:
[RFC] list_stored_sd_metadata API. (#160610)

Summary:

1\ Certain checkpoint load use cases are not aware of the properties of the data/tensors they want to load.
2\ These usecases include data loader checkpoints, reading data for post processing (when the original model definition is not available).
3\ There, we have to use saved checkpoint  (metadata) as our source of truth.
4\ This RFC proposal exposes the checkpoint metadata using a public API.

In this proposal we expose the stored state-dict metadata  (minus associated storage/chunk metadata). 

Chunk/storage details should not be exposed to the users and is a impl detail of the storage writer/reader.

Test Plan:
buck test 'fbcode//mode/dev-nosan' fbcode//caffe2/test/distributed/checkpoint:test_list_stored_state_dict


lintrunner --force-color --tee-json=lint.json --skip CLANGTIDY,CLANGFORMAT,MYPY,MYPYSTRICT torch/distributed/checkpoint/state_dict_loader.py

Reviewed By: saumishr

Differential Revision: D80231457

pradeepfn commented:
@pytorchbot merge

jeffdaily commented:
New unit test introduced in this PR is causing ROCm failures.

https://hud.pytorch.org/failure?name=periodic%20%2F%20linux-jammy-rocm-py3.10%20%2F%20test%20(distributed%2C%202%2C%203%2C%20linux.rocm.gpu.mi250.4%2C%20module%3Arocm%2C%20oncall%3Adistributed)&jobName=undefined&failureCaptures=distributed%2Fcheckpoint%2Ftest_list_stored_state_dict.py%3A%3ATestListStateDict%3A%3Atest_list_stored_sd_metadata

jeffdaily commented:
@pytorchbot revert -c nosignal -m "broke ROCm CI, but flaky also on CUDA CI https://hud.pytorch.org/failure?name=periodic%20%2F%20linux-jammy-rocm-py3.10%20%2F%20test%20(distributed%2C%202%2C%203%2C%20linux.rocm.gpu.mi250.4%2C%20module%3Arocm%2C%20oncall%3Adistributed)&jobName=undefined&failureCaptures=distributed%2Fcheckpoint%2Ftest_list_stored_state_dict.py%3A%3ATestListStateDict%3A%3Atest_list_stored_sd_metadata"
